### PR TITLE
CRM-19129 Allow Users to Select the From Address used for when sendin…

### PR DESCRIPTION
**Pull request by seamuslee001**
…g Email Receipts

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
test this please

File: CRM/Contribute/Form/Task/PDF.php [hunk 1]
```diff
@@ -187,6 +195,14 @@ public function postProcess() {
       $objects['contribution']->receive_date = CRM_Utils_Date::isoToMysql($objects['contribution']->receive_date);
 
       $values = array();
+      if (isset($params['fromEmailAddress']) && !$elements['createPdf']) {
+        $fromEmail = $params['fromEmailAddress'];
+        $from = CRM_Utils_Array::value($fromEmail, $this->_emails);
+        $fromDetails = explode(' <', $from);
```
Comment (jitendrapurohit) [inline CRM/Contribute/Form/Task/PDF.php]: Can we avoid these lines to be re-written ?

Comment (seamuslee001) [inline CRM/Contribute/Form/Task/PDF.php]: The problem is when you load the from addresses from the optionlist it in the format of "Name" <email>" and there are array keys one for the name and one for the email in the receipt so there has to be this conversion

Comment (seamuslee001) [inline CRM/Contribute/Form/Task/PDF.php]: Also $params['fromEmailAddress'] outputs as the value id for the relevant option value when you select that is from the list of from email addresses

Comment (jitendrapurohit) [inline CRM/Contribute/Form/Task/PDF.php]: No, I meant, as we are already converting and storing `$fromDetails` at line 157-159. Do we require to write these lines again? As I find this to be working if I remove the above three lines.

Comment (seamuslee001) [inline CRM/Contribute/Form/Task/PDF.php]: Oh by bad woops thanks @jitendrapurohit

**Comment by seamuslee001** [comment]
@jitendrapurohit I think this has fixed the issue where if no emails are found on the logged in contact's record. If you can test again that would be great

File: tests/phpunit/CRM/Contact/Form/Task/EmailCommonTest.php [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+ | see the CiviCRM license FAQ at http://civicrm.org/licensing        |
+ +--------------------------------------------------------------------+
+ */
+ /**
+  * Test class for CRM_Contact_Form_Task_EmailCommon.
+  * @group headless
+  */
+class CRM_Contact_Form_Task_EmailCommonTest extends CiviUnitTestCase {
+
+  protected function setUp() {
```
Comment (eileenmcnaughton) [inline tests/phpunit/CRM/Contact/Form/Task/EmailCommonTest.php]: what a beautiful test :-)

File: CRM/Contact/Form/Task/EmailCommon.php [hunk 3]
```diff
@@ -88,43 +104,48 @@ public static function preProcessFromAddress(&$form) {
...
+    foreach ($form->_fromEmails as $key => $fromEmail) {
+      if (empty($fromEmail)) {
+        unset($form->_fromEmails[$key]);
+      }
     }
-    if (!empty($defaultEmail['signature_text'])) {
-      $defaults['text_message'] = "\n\n--\n" . $defaultEmail['signature_text'];
+    if (!empty($emails)) {
+      // Add signature
+      $defaultEmail = civicrm_api3('email', 'getsingle', array('id' => key($form->_fromEmails)));
```
Comment (jitendrapurohit) [inline CRM/Contact/Form/Task/EmailCommon.php]: This throws an exception(check JIRA for reproducing) when `key($form->_fromEmails)` evaluates to `FALSE`. As key() returns the index element of the **current** array position, this current pointer [return a boolean FALSE](http://php.net/manual/en/function.current.php) after the array is being looped.

Hence, as the foreach() just targets to remove the empty values from an array, replacing it with an array functions should keep the pointer unaffected. Something like -

    $form->_fromEmails = array_filter($form->_fromEmails);

OR `reset($form->_fromEmails)` after the loop. Checking for integer value before calling the api would avoid additional exceptions to be thrown as domainEmails() function return string as a key.

Comment (seamuslee001) [inline CRM/Contact/Form/Task/EmailCommon.php]: Thanks should be fixed in latest push

**Comment by jitendrapurohit** [comment]
All noted issues seems to be fixed. Thanks for the quick update @seamuslee001 :-). This can be tagged as merge on pass.

**Pull request merged**